src/main/scala/cognite/spark/v1/DataModelInstancesRelation.scala

```diff
@@ -179,99 +165,104 @@ class DataModelInstanceRelation(config: RelationConfig, modelExternalId: String)
+    case x: Array[Long] => ArrayProperty(x.toVector.map(Int64Property))
+    case x: Array[String] => ArrayProperty(x.toVector.map(StringProperty))
+    case x: Array[Boolean] => ArrayProperty(x.toVector.map(BooleanProperty))
+    case x: Array[java.math.BigDecimal] =>
+      ArrayProperty(x.toVector.map(i => Float64Property(i.doubleValue())))
+    case x: Array[java.math.BigInteger] =>
+      ArrayProperty(x.toVector.map(i => Int64Property(i.longValue())))
+    case x: LocalDate => DateProperty(x)
+    case x: java.sql.Date => DateProperty(x.toLocalDate)
+    case x: LocalDateTime => DateProperty(x.toLocalDate)
+    case x: Instant => TimeStampProperty(x.atZone(ZoneId.systemDefault()))
+    case x: java.sql.Timestamp =>
+      TimeStampProperty(x.toInstant.atZone(ZoneId.systemDefault()))
+    case x => throw new CdfSparkException(s"Cannot parse the value with udentified type: $x")
   }
 
-  private def parseFromJson(value: Json, propType: String): Option[Any] =
-    if (value.isNull) {
-      Some(null) // scalastyle:off null
-    } else {
-      propType.toLowerCase match {
-        case "text" =>
-          value.asString
-        case "boolean" =>
-          value.asBoolean
-        case "numeric" | "float64" =>
-          value.asNumber.map(_.toDouble)
-        case "float32" =>
-          value.asNumber.map(_.toFloat)
-        case "int" | "int32" =>
-          value.asNumber.flatMap(_.toInt)
-        case "int64" | "bigint" =>
-          value.asNumber.flatMap(_.toLong)
-        case "text[]" =>
-          Some(value.asArray.getOrElse(Vector()).flatMap(_.asString))
-        case "boolean[]" =>
-          Some(value.asArray.getOrElse(Vector()).flatMap(_.asBoolean))
-        case "numeric[]" | "float64[]" =>
-          Some(value.asArray.getOrElse(Vector()).flatMap(_.asNumber.map(_.toDouble)))
-        case "float32[]" =>
-          Some(value.asArray.getOrElse(Vector()).flatMap(_.asNumber.map(_.toFloat)))
-        case "int[]" | "int32[]" =>
-          Some(value.asArray.getOrElse(Vector()).flatMap(_.asNumber.flatMap(_.toInt).toList))
-        case "int64[]" | "bigint[]" =>
-          Some(value.asArray.getOrElse(Vector()).flatMap(_.asNumber.flatMap(_.toLong).toList))
-        case a => throw new CdfSparkException(unknownPropertyTypeMessage(a))
-      }
-    }
+  private def fromPropertyType(x: PropertyType): Any = x match {
+    case Int32Property(value) => value
+    case Int64Property(value) => value
+    case Float32Property(value) => value
+    case Float64Property(value) => value
+    case BooleanProperty(value) => value
+    case StringProperty(value) => value
+    case DateProperty(value) => value
+    case TimeStampProperty(value) => value
+    case DirectRelationProperty(value) => value
+    case GeographyProperty(value) => value
+    case GeometryProperty(value) => value
+    case ArrayProperty(values) => values.map(fromPropertyType)
+    case x => throw new CdfSparkException(s"Unknown property type with value $x")
```

Comment:
Similarly, `s"Unknown property type ${x.getClass.getName} with value ${x.toString}"` could be more helpful.